libs/vkd3d/command.c

```diff
@@ -3474,87 +3474,231 @@ static HRESULT d3d12_command_list_allocate_transfer_buffer(struct d3d12_command_
+        extent.depth = dst_view_desc.layer_count;
+
+        if (!d3d12_command_list_create_framebuffer(list, pipeline_info.vk_render_pass, 1, &dst_view->u.vk_image_view, extent, &vk_framebuffer))
+        {
+            ERR("Failed to create framebuffer.\n");
+            goto cleanup;
+        }
+
+        begin_info.sType = VK_STRUCTURE_TYPE_RENDER_PASS_BEGIN_INFO;
+        begin_info.pNext = NULL;
+        begin_info.renderPass = pipeline_info.vk_render_pass;
+        begin_info.framebuffer = vk_framebuffer;
+        begin_info.clearValueCount = 0;
+        begin_info.pClearValues = NULL;
+        begin_info.renderArea.offset.x = 0;
+        begin_info.renderArea.offset.y = 0;
+        begin_info.renderArea.extent.width = extent.width;
+        begin_info.renderArea.extent.height = extent.height;
+
+        viewport.x = (float)region->dstOffset.x;
+        viewport.y = (float)region->dstOffset.y;
+        viewport.width = (float)region->extent.width;
+        viewport.height = (float)region->extent.height;
+        viewport.minDepth = 0.0f;
+        viewport.maxDepth = 1.0f;
+
+        scissor.offset.x = region->dstOffset.x;
+        scissor.offset.y = region->dstOffset.y;
+        scissor.extent.width = region->extent.width;
+        scissor.extent.height = region->extent.height;
+
+        push_args.offset.x = region->srcOffset.x - region->dstOffset.x;
+        push_args.offset.y = region->srcOffset.y - region->dstOffset.y;
+
+        vk_descriptor_set = d3d12_command_allocator_allocate_descriptor_set(
+                list->allocator, pipeline_info.vk_set_layout,
+                VKD3D_DESCRIPTOR_POOL_TYPE_STATIC);
+
+        if (!vk_descriptor_set)
+        {
+            ERR("Failed to allocate descriptor set.\n");
+            goto cleanup;
+        }
+
+        vk_image_info.sampler = VK_NULL_HANDLE;
+        vk_image_info.imageView = src_view->u.vk_image_view;
+        vk_image_info.imageLayout = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+
+        vk_descriptor_write.sType = VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET;
+        vk_descriptor_write.pNext = NULL;
+        vk_descriptor_write.dstSet = vk_descriptor_set;
+        vk_descriptor_write.dstBinding = 0;
+        vk_descriptor_write.dstArrayElement = 0;
+        vk_descriptor_write.descriptorCount = 1;
+        vk_descriptor_write.descriptorType = VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE;
+        vk_descriptor_write.pImageInfo = &vk_image_info;
+        vk_descriptor_write.pBufferInfo = NULL;
+        vk_descriptor_write.pTexelBufferView = NULL;
+
+        VK_CALL(vkUpdateDescriptorSets(list->device->vk_device, 1, &vk_descriptor_write, 0, NULL));
+
+        for (i = 0; i < ARRAY_SIZE(vk_image_barriers); i++)
+        {
+            vk_image_barriers[i].sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
+            vk_image_barriers[i].pNext = NULL;
+            vk_image_barriers[i].srcAccessMask = 0;
```

Comment:
Shouldn't this be VK_ACCESS_COLOR/DEPTH WRITE_BIT?

Reply:
The first barrier only performs a layout transition so we shouldn't need it; we set the `*_ATTACHMENT_WRITE_BIT` for the second barrier.

Reply:
I think we need to set ATTACHMENT_WRITE_BIT after the layout transition into ATTACHMENT_OPTIMAL. Memory needs to be made visible.

Reply:
FWIW, this is dstAccessMask

Reply:
In order to make it visibl we'd have to put it in `dstStageMask` though, and as far as I'm aware, putting write bits there has no effect.

Do we need the `READ` bits? We don't use blending, our depth test always passes and we don't do anything with stencil.

Reply:
If the renderpass is created with LOAD, then it accesses COLOR_ATTACHMENT_READ. For copy, we should probably use a DONT_CARE render pass.

Reply:
We're currently doing a TRANSFER -> FRAGMENT execution barrier here, with no dstAccessMask after the layout transition. I believe WRITE must be in dstAccessMask to make sure the writes are ordered (and it cannot write to a stale FB cache or something), but it's a corner case I haven't explored all that much. I'm aware WAR does not require srcAccessMask to be set, but not so sure about dstAccessMask when layout transitions are in the picture.

Reply:
The question is if memory must be visible if the only thing that happens are write operations.

Reply:
Added both the read and write bits to be on the safe side, and also fixed that we weren't resetting the `dstAccessMask` for the second barrier.